artifact/v3/manifest/manifest.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <artifact/v3/manifest/manifest.hpp>
+
+#include <string>
+#include <sstream>
+
+#include <common/error.hpp>
+
+namespace mender {
+namespace artifact {
+namespace v3 {
+namespace manifest {
+
+namespace io = mender::common::io;
+namespace error = mender::common::error;
+
+const size_t expected_shasum_length {64};
+const size_t expected_whitespace {2};
+
+ExpectedManifest Parse(mender::common::io::Reader &reader) {
+	Manifest m {};
+
+	stringstream ss {};
+
+	io::StreamWriter sw {ss};
+
+	auto err = io::Copy(sw, reader);
+	if (err) {
+		return expected::unexpected(err);
+	}
+
+	string entry_name {};
+	string shasum {};
+	string line {};
+
+	while (getline(ss, line, '\n')) {
```

Comment:
Ole, forgive me, no offense to you, this is nothing personal and I know this is how we decided to do things.

But the above is really just crazy to me. We want to read lines from a file. Our answer is to:
1. create a file stream
2. create a Reader on top of the stream
3. create a Writer on top of a string stream
4. use Copy to shovel data between the Reader and Writer and thus from one stream to the other
5. get lines from the string stream

I don't even want to count the buffers used in the above, but if our target is to limit the amount of our code and (re)use what's available, why are we doing this to ourselves?

Reply:
Plus the fact that a large input file will just kill any process running this code if I understand correctly what's actually happening in those 5 steps above.

Reply:
> because we still internally build up a JSON structure 

AFAICT, here we don't build any JSON structure, we just read lines from a file.

Reply:
> getline requires a stream instead of iterators or a buffer. This is easily solvable by implementing the inverse of [StreamReader](https://github.com/mendersoftware/mender/blob/feature-c%2B%2B-client/common/io.hpp#L108).

So we would have `StreamReader` wrapping a stream into a `Reader` and then `ReaderStream` which would provide a stream for a given `Reader`? That sounds twisted to me.

Reply:
I believe the general issue is that the Reader/Writer concept is not native to C++. C++ has streams, `istream` for input and `ostream` for output. Libraries and even the standard library are implemented the way that they work with streams. Why should we introduce the Reader/Writer concept into that and break/disrupt things? What is the benefit? AFAICT, everywhere where we have a `Reader` an `istream` would do and the same applies to `Writer` and `ostream`. Then we would be able to stream-process data and use the C++-native APIs without any extra code. Instead of wrapping a stream into our class and then providing the stream API on top of it, we could just work with the streams directly. C++ already has an abstraction for I/O, there's no need to introduce another one, IMHO.

The question is how hard it is to provide a stream on top of a buffer-based API, like the one that Beast or tiny-process-lib provide. That I will try to find out.

Also, this discussion is unrelated to this PR, we might as well take it somewhere else.

Reply:
Correct me if I'm wrong, but AFAIK the C++ streams are not virtual, and therefore they are unsuitable for interfaces. I believe they also require a lot more methods to be implemented to subclass them. But feel free to suggest improvements which are easier to use than Reader and Writer.

Reply:
I am going to close this for now. This can be handled out of band for this PR.